src/firefly/js/visualize/draw/ShapeDataObj.js

```diff
@@ -605,16 +605,40 @@ function drawLine(drawObj, ctx,  plot, drawParams, onlyAddToPath) {
         ctx.moveTo(devPt0.x, devPt0.y);
         ctx.lineTo(devPt1.x, devPt1.y);
         if (!onlyAddToPath || style===Style.HANDLED) DrawUtil.stroke(ctx);
+        if (withArrow) {
+            DrawUtil.drawArrowOnLine(ctx, devPt0, devPt1, color);
+        }
     }
 
     if (!isNil(text) && inView) {
-        const textLocPt= makeTextLocationLine(plot, textLoc, fontSize,pts[0], pts[1]);
-        drawText(drawObj,  ctx, plot, plot.getDeviceCoords(textLocPt), drawParams);
+        if (textLoc === TextLocation.LINE_TOP_STACK) {
+            const textLines = text.split('\n');
+            textLines.forEach((oneText, idx) => {
+                const tLocPt = makeTextLocationLine(plot, textLoc, fontSize, pts[0], pts[1], idx+1, drawObj);
+                drawObj.text = oneText;
+                drawText(drawObj,  ctx, plot, plot.getDeviceCoords(tLocPt), drawParams);
+            });
+            drawObj.text = text;
+        } else {
+            const textLocPt = makeTextLocationLine(plot, textLoc, fontSize, pts[0], pts[1]);
+            drawText(drawObj, ctx, plot, plot.getDeviceCoords(textLocPt), drawParams);
+        }
     }
 
-    if (style===Style.HANDLED && inView) {
-        DrawUtil.fillRec(ctx, color,devPt0.x-2, devPt0.y-2, 5,5, renderOptions);
-        DrawUtil.fillRec(ctx, color,devPt1.x-2, devPt1.y-2, 5,5, renderOptions);
+    if ([Style.HANDLED, Style.STARTHANDLED, Style.ENDHANDLED].includes(style) && inView) {
+        const rAngle = VisUtil.computeSimpleSlopeAngle(devPt0, devPt1);
+        const rOptions = {};
+
+        rOptions.rotAngle = rAngle;
+
+        if (style !== Style.ENDHANDLED) {
+            rOptions.translation = {x: devPt0.x, y: devPt0.y};
+            DrawUtil.fillRec(ctx, color, -3, -3, 7, 7, rOptions);
+        }
+        if (style !== Style.STARTHANDLED) {
+            rOptions.translation = {x: devPt1.x, y: devPt1.y};
+            DrawUtil.fillRec(ctx, color, -3, -3, 7, 7, rOptions);
+        }
```

Comment:
I am having trouble understanding this logic. It looks like if the is `HANDLED` it does stuff twice.  I think is would be better to say:
```js
if (style === Style.HANDLED) {
   // do this
}
else if (style === Style.ENDHANDLED) {
   // do this
}
else  if (style === Style.STARTHANDLED) {
   // do this
}
```